### ISSUE TITLE
"Feb 2011" parsing fails but Jan, Mar-Dec works

### ISSUE THREAD

**Issue by umrashrf** [dateutil/dateutil#25]
```python
>>> from dateutil import parser
>>> parser.parse('Jan 2011', fuzzy=True)
datetime.datetime(2011, 1, 30, 0, 0)
>>> parser.parse('Feb 2011', fuzzy=True)
Traceback (most recent call last):
  File "<console>", line 1, in <module>
  File "/home/umair/dev/venvs/nathanartz/local/lib/python2.7/site-packages/dateutil/parser.py", line 743, in parse
    return DEFAULTPARSER.parse(timestr, **kwargs)
  File "/home/umair/dev/venvs/nathanartz/local/lib/python2.7/site-packages/dateutil/parser.py", line 310, in parse
    ret = default.replace(**repl)
ValueError: day is out of range for month
>>> parser.parse('Dec 2011', fuzzy=True)
datetime.datetime(2011, 12, 30, 0, 0)
```

**Comment by Allactaga** [comment]
Actually it is not only problem with February. I believe tomorrow (Jan 31st) would bring even more cases like this: 
```python
>>> from datetime import date
>>> from dateutil import parser
>>> parser.parse("Apr 2011", default=date(2015, 1, 31))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/dateutil/parser.py", line 697, in parse
    return DEFAULTPARSER.parse(timestr, **kwargs)
  File "/System/Library/Frameworks/Python.framework/Versions/2.7/Extras/lib/python/dateutil/parser.py", line 310, in parse
    ret = default.replace(**repl)
ValueError: day is out of range for month
```

This is happening because case that there could be less days in the replacing month, than there are days in the default date, should be checked before [this line](https://github.com/dateutil/dateutil/blob/master/dateutil/parser.py#L310).

**Comment by umrashrf** [comment]
Right. If this is coming from default date then I think day should come from **calendar** depending on month.

**Comment by pganssle** [comment]
Presumably the correct solution to this is to interpret `"Feb 2011"` as the first day of February. By extension, is it reasonable to say that `"2011"` should be interpreted as 0:00:00 January 1st, 2011? I'm getting the impression that the parser falls back to modifying the current date rather than falling back to the first instance of a date that can be described by the parsed information.

I think the problem is that there are two concepts - a date-time stamp refers to a *span of time*, and can be reasonably specified within context. Generally, if an element is left to context, then if the element is *smaller* than the smallest specified element, you use the first point of the span in that dimension. If an element is *larger* than the largest specified element, you use the *most recent* time which can be reasonably be described by the date-time (e.g. "February" parses to "February 1, 2014 0:00:00"). Generally in common parlance, you specify every element between the largest and smallest elements ("February 2014 3:00 PM" is generally considered ambiguous).

`parser` seems to try to approximate the "Use the most recent" by just modifying the current date-stamp, which defeats peoples' expectations. I suggest that we explicitly modify the parser to use the scheme I detailed above. For meaningless stuff like "February 2014 3:00 PM", we'll fall back separately on dates and on times - so that will refer to February 1, 2014 15:00:00.

It might be worthwhile to do this in a beta or alpha release and solicit feedback from major users of dateutil, since it potentially breaks the API if someone is relying on the current (albeit undocumented) fallback behavior.

**Comment by umrashrf** [comment]
My preference would be to use last day of each month because it's easier to replace the resulting date to have first day of each month.
```python
>>> from dateutil import parser
>>>
>>> parser.parse('Feb 2011')
datetime.datetime(2011, 2, 28, 0, 0)
>>> 
>>> parser.parse('Feb 2011').replace(day=1)
datetime.datetime(2011, 1, 1, 0, 0)
```

**Comment by pganssle** [comment]
@umrashrf That's a good point, but I think it's a bit counter-intuitive in terms of expected behavior, plus it just pushes the burden of determing what the last day of the month is into dateutil. That's not necessarily bad, but I feel like having the default date be the end of the month is such an unlikely use-case that it's probably not the best way to go; if people were equally likely to want the first or last day of the month, I'd say your way is better, but my hunch is that -- if forced to chose -- 90%+ would want it to be the first day of the month.

I've started working on how this is implemented, I'm thinking that the best thing to do might be to default to the first day of the month, but have a flag or combination of inputs that can be specified to let users customize the behavior of the parser.

**Comment by pganssle** [comment]
@umrashrf @Allactaga As I'm documenting this code, I realize that there's actually a simple workaround for this bug. Simply specify a `default` datetime. Currently, if `default` is `None`, it uses midnight at the current date and replaces whatever elements are specified. In the intermediate term, or when using old versions of dateutil, you can solve the problem by specifying the first of the month:

    from datetime import datetime
    from dateutil import parser
    
    datedflt = datetime.now().replace(day=1, hour=0, minute=0, second=0, microsecond=0)
    date_out = parser.parse("Feb 2011", default=datedflt)

Hopefully this tides you over while a more permanent solution is crafted.

**Comment by Allactaga** [comment]
Hi @pganssle , this solution would require knowledge about the parsing date. Please consider '12:15' as a given string.

**Comment by pganssle** [comment]
@Allactaga Yes, agreed, that's why I'm not suggesting patching the parser like that. I was assuming that you knew that you had something of the form "Month Year".  Another option if you happen to know that *sometimes* you'll have dates like that is to do it this way (and again, this is a workaround, a real solution is definitely forthcoming):

```python
from dateutil import parser
from datetime import datetime
try:
    date_out = parser.parse("Feb 2011", fuzzy=True)
except ValueError:
    # If the ValueError was caused because the date in the default date is out of range, try it again with a new default
    defaultdate = datetime.now().replace(day=1, hour=0, minute=0, second=0, microsecond=0)
    date_out = parser.parse("Feb 2011", fuzzy=True, default=defaultdate)
```

Parsing "12:15" won't raise an exception, and anything other than this problem that would raise an exception the first time around should still raise the exception the second time around. Hopefully this can work in the meantime (and if you want to support a lower dependency level of dateutil than the first release in which this bug is fixed).

**Comment by pganssle** [comment]
Fixed with #30
